### PR TITLE
Add support for basic QoS and pod level cgroup management

### PR DESCRIPTION
**Pull request by dubstack**
This PR is a WIP and is tied to this upstream issue #27204 
It adds support for creation,deletion and updates of cgroups in Kubernetes.
@vishh PTAL

Please note that the first commit is part of this PR: #27749
cc @kubernetes/sig-node

Signed-off-by: Buddha Prakash buddhap@google.com

### CODE REVIEW

File: pkg/kubelet/cm/types.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cm
+
+// ResourceConfig holds information about all the supported cgroup resource parameters.
+type ResourceConfig struct {
+	// Memory limit (in bytes).
+	Memory int64
```
Comment (vishh) [inline pkg/kubelet/cm/types.go]: nit: Make this a pointer to help deal with scenarios where memory limits need not be set.

File: pkg/kubelet/cm/types.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+// CgroupManager is an interface which defines the manager which allows for cgroup management.
+// Supports Cgroup Creation , Deletion and Updates.
+type CgroupManager interface {
+	//TODO Create and update should ideally return the created cgroup's
+	// fully qualified name, but first we would need to fix
+	// what "Name" these methods take as Config.
+	// as systemd and cgroupfs would expect different inputs
+
+	// Creates and applies the cgroup configurations on the cgroup.
+	Create() error
```
Comment (vishh) [inline pkg/kubelet/cm/types.go]: nit: Why is Create not taking in `CgroupConfig`?

File: pkg/kubelet/cm/types.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+	// TODO Needs to be defined exactly
+	// Fully qualified name
+	Name string
+	// ResourceParameters contains various cgroups settings to apply.
+	ResourceParameters *ResourceConfig
+}
+
+// CgroupManager is an interface which defines the manager which allows for cgroup management.
+// Supports Cgroup Creation , Deletion and Updates.
+type CgroupManager interface {
```
Comment (vishh) [inline pkg/kubelet/cm/types.go]: I'd expect `CgroupManager` to be a global utility for creating and deleting multiple cgroups.

File: pkg/kubelet/cm/cgroup_manager.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	// Libcontainer raw fs cgroup manager.
+	fsCgroupManager libcontainercgroups.Manager
+}
+
+// Make sure that cgroupManagerImpl implements the CgroupManager interface
+var _ CgroupManager = &cgroupManagerImpl{}
+
+// NewCgroupManagerImpl is a factory method that returns
+// a configured CgroupManager
+func NewCgroupManagerImpl(cgroupConfig *CgroupConfig) (CgroupManager, error) {
```
Comment (vishh) [inline pkg/kubelet/cm/cgroup_manager.go]: I don't think we need a cgroup manager per container. We only need one instance of cgroup manager to manage lifecycle of all cgroups..

File: pkg/kubelet/cm/cgroup_manager.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+	// the containers absolute name
+	name := cgroupConfig.Name
+	libcontainerCgroupConfig := getLibcontainerCgroupConfig(cgroupConfig)
+	// get all the mounted subsystems
+	cgroupSubsystems, err := getCgroupSubsystems()
+	if err != nil {
+		return &cgroupManagerImpl{}, fmt.Errorf("failed to get cgroup subsystems: %v", err)
+	}
+	// Get map of all cgroup paths on the system
+	cgroupPaths := make(map[string]string, len(cgroupSubsystems.mountPoints))
```
Comment (vishh) [inline pkg/kubelet/cm/cgroup_manager.go]: Why is this necessary?

**Comment by vishh** [comment]
ok to test

**Comment by vishh** [comment]
@dubstack Finished my first pass. Ping me once you address the comments.

File: pkg/kubelet/cm/types.go (hunk 3 above)
Comment (dubstack) [inline pkg/kubelet/cm/types.go]: Okay. I thought we will have one manager per cgroup. Thats why no cgroupConfig in Create(). I will go ahead and make changes.

File: pkg/kubelet/cm/types.go (hunk 1 above)
Comment (dubstack) [inline pkg/kubelet/cm/types.go]: Ack!

File: pkg/kubelet/cm/cgroup_manager.go (hunk 5 above)
Comment (dubstack) [inline pkg/kubelet/cm/cgroup_manager.go]: We need this for getting the cgroup paths that exist for the particular cgroup. This is needed for us to destroy the cgroup. In the case of global cgroup manager for all cgroup management, It will be clearer once I make the changes.

**Comment by dubstack** [comment]
@vishh Addressed your comments. PTAL. I think we can't really test much at this point. So shifting the testing to the next PR.

File: pkg/kubelet/cm/helpers.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+		return &cgroupfs.Manager{}, fmt.Errorf("Unable to destroy cgroup paths for cgroup %v: %v", name, err)
+	}
+	// Get map of all cgroup paths on the system for the particular cgroup
+	cgroupPaths := make(map[string]string, len(cgroupSubsystems.mountPoints))
+	for key, val := range cgroupSubsystems.mountPoints {
+		cgroupPaths[key] = path.Join(val, name)
+	}
+	return &cgroupfs.Manager{
+		Cgroups: libcontainerCgroupConfig,
+		Paths:   cgroupPaths,
```
Comment (vishh) [inline pkg/kubelet/cm/helpers.go]: AFAIK paths are not required. Take a look at the logic in `container_manager_linux.go` for reference.

File: pkg/kubelet/cm/types.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+// cgroup information to both systemd and raw cgroup fs
+// implementation of the Cgroup Manager interface.
+type CgroupConfig struct {
+	// We would expect systemd implementation to make appropriate
+	// name conversion. For example, if we pass /foo/bar
+	// then systemd should convert the name to something like
+	// foo.slice/foo-bar.slice
+
+	// Fully qualified name
+	Name string
```
Comment (vishh) [inline pkg/kubelet/cm/types.go]: As part of the interface will you just create the leaf cgroups or the whole tree?

**Comment by vishh** [comment]
Two comments, otherwise lgtm

File: pkg/kubelet/cm/helpers.go (hunk 6 above)
Comment (dubstack) [inline pkg/kubelet/cm/helpers.go]: We need it here only for the destroy() method. We create a new cgroupfs manager object for each method. Because of this when destory() method is called Paths need to be initialized.

I believe in other places the same object  which created the cgroup must be used to destroy the cgroups. In that case it paths is stored when the apply function is called.

File: pkg/kubelet/cm/types.go (hunk 7 above)
Comment (dubstack) [inline pkg/kubelet/cm/types.go]: I was thinking of just creating the leaf cgroups. I assume that we wont be passing paths where the parent directory doesn't already exist. Does that sound good?

Comment (vishh) [inline pkg/kubelet/cm/types.go]: Yeah. That's fine. But mention that behavior explicitly.

Comment (dubstack) [inline pkg/kubelet/cm/types.go]: Ack! Thanks.

**Comment by dubstack** [comment]
@vishh PTAL. I think paths would be required for us to destroy the cgroups. Does that sound reasonable?

File: pkg/kubelet/cm/helpers.go [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+	// Cgroup subsystem to their mount location.
+	// e.g.: "cpu" -> "/sys/fs/cgroup/cpu"
+	mountPoints map[string]string
+}
+
+// GetCgroupSubsystems returns information about the mounted cgroup subsystems
+func getCgroupSubsystems() (cgroupSubsystems, error) {
+	// Get all cgroup mounts.
+	allCgroups, err := libcontainercgroups.GetCgroupMounts()
```
Comment (vishh) [inline pkg/kubelet/cm/helpers.go]: Can we cache the mounts?

File: pkg/kubelet/cm/helpers.go (hunk 6 above)
Comment (vishh) [inline pkg/kubelet/cm/helpers.go]: Ack. Let's cache it

**Comment by dubstack** [comment]
@vishh I made suggested changes. I also realized that this PR could be made independent of other PR's. So can we get this merged aswell.?

File: pkg/kubelet/cm/cgroup_manager_linux.go [hunk 9]
```diff
@@ -0,0 +1,115 @@
+// +build linux
```
Comment (vishh) [inline pkg/kubelet/cm/cgroup_manager_linux.go]: This tag is not required.

File: pkg/kubelet/cm/cgroup_manager_unsupported.go [hunk 10]
```diff
@@ -0,0 +1,42 @@
+// +build !linux
```
Comment (vishh) [inline pkg/kubelet/cm/cgroup_manager_unsupported.go]: This tag is not required.

File: pkg/kubelet/cm/helpers.go [hunk 11]
```diff
@@ -0,0 +1,100 @@
+// +build linux
```
Comment (vishh) [inline pkg/kubelet/cm/helpers.go]: This should also be renamed to `helpers_linux.go` right?

Comment (dubstack) [inline pkg/kubelet/cm/helpers.go]: Ack!

File: pkg/kubelet/cm/cgroup_manager_unsupported.go (hunk 10 above)
Comment (dubstack) [inline pkg/kubelet/cm/cgroup_manager_unsupported.go]: Ack!

File: pkg/kubelet/cm/helpers.go (hunk 11 above)
Comment (dubstack) [inline pkg/kubelet/cm/helpers.go]: Ack!

**Comment by vishh** [comment]
LGTM except for the pending comments.

**Comment by vishh** [comment]
LGTM

**Comment by dubstack** [comment]
@k8s-bot e2e test this issue #27296

**Comment by dubstack** [comment]
@k8s-bot e2e test this issue #27296

**Comment by dubstack** [comment]
@k8s-bot e2e test this issue #27462

**Comment by dubstack** [comment]
@k8s-bot node e2e test this issue #27462

**Comment by dubstack** [comment]
@k8s-bot node e2e test this issue  #27296

**Comment by dubstack** [comment]
@k8s-bot e2e test this

**Comment by dubstack** [comment]
Applied LGTM as it was removed by mistake.

**Comment by liggitt** [comment]
cgroup_manager_unsupported.go references a type defined in helpers_linux.go, which breaks compiling on non-linux platforms

**Comment by dubstack** [comment]
Oh no! Thanks @liggitt for reporting this.

**Comment by jlowdermilk** [comment]
So it looks like this change also broke a pvc unit test, though I'm not sure how that could be. 

See https://console.cloud.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-test-go/13556, which started failing with the commit that merged this change
```
Error Message

Failed
Stacktrace

framework_test.go:810: Test "5-6 - update bound volume from 1.2": Claim check failed [A-expected, B-got result]: {"claim5-6":{"metadata":{"name":"claim5-6","namespace":"default","selfLink":"/api/v1/pvc/claim5-6","uid":"uid5-6","creationTimestamp":null
	
	A: ,"annotations":{"pv.kubernetes.io/bind-completed":"yes"}},"spec":{"accessModes":["ReadWriteOnce","ReadOnlyMany"],"resources":{"requests":{"storage":"1Gi"}},"volumeName":"volume5-6"},"status":{"phase":"Bound"}}}
	
	B: },"spec":{"accessModes":["ReadWriteOnce","ReadOnlyMany"],"resources":{"requests":{"storage":"1Gi"}},"volumeName":"volume5-6"},"status":{"phase":"Bound"}}}
```

**Comment by dubstack** [comment]
Currently This PR is not being used in other parts of the codebase. So I am not sure why it would break any tests.

**Comment by jlowdermilk** [comment]
Re the pvc test, it's now passing again with the same commit, so I suspect framework problems. Will look into elsewhere.

**Comment by dubstack** [comment]
That makes sense. I was having troubles with the same unit test error aswell before this commit.
